src/rust/src/lib.rs

```diff
@@ -1,132 +1,224 @@
 use extendr_api::prelude::*;
-use std::collections::HashMap;
+use std::iter::zip;
 
 #[extendr]
 fn model_matrix(data: List) -> Result<Robj> {
     let nrow = data.iter().next().map(|(_, col)| col.len()).unwrap_or(0);
-    let mut processed_columns: Vec<Array2<f64>> = Vec::new();
-    let mut column_names: Vec<String> = Vec::new();
+    let columns = data.iter().map(|(id, col)| {
+        if col.is_string() {
+            (id, R!("factor({{col}})").unwrap())
+        } else {
+            (id, col)
+        }
+    });
```

Comment:
Manually converting strings (`character` vector) to factors is a substantial challenge. R has its string interner, and it is highly optimised, so I let R do the conversion instead.